File: src/com/rentalbikes/CompositeFamilyPromotion.java
package com.rentalbikes;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

public class CompositeFamilyPromotion implements iRentable {

	private static int minPromos = 3;
	private static int maxPromos = 5;

	private static int discount = 30;
	
	private List<iRentable> promoRentals = new ArrayList<iRentable>();

	
	/**
	 * 
	 */
	@Override
	public BigDecimal getTotalCost() {
		if (this.promoRentals.size() < this.minPromos) {
			return new BigDecimal(0);
		}
		
		long promoSums = 0;
		
		for (iRentable rental : this.promoRentals) {
			promoSums = promoSums + rental.getTotalCost().longValueExact();
		} 		
		return new BigDecimal(promoSums * discount / 100);
	}
	
	/**
	 * 
	 * @param aRental
	 * @throws Exception
	 */
	public void addRental(iRentable aRental) throws Exception {
		if (this.promoRentals.size() < this.maxPromos) {
			this.promoRentals.add(aRental);
		}else {
			throw new Exception("Maximum rental items reached");
		}
	}
}

File: src_python/CompositeFamilyPromotion.py
from AbstractRentable import AbstractRentable
import unittest
from RentByDay import RentByDay

class CompositeFamilyPromotion(AbstractRentable):
    def __init__(self):
        self.minPromos = 3
        self.maxPromos = 5
        self.discount = 30
        self.promoRentals = list()

    def getTotalCost(self):
        if len(self.promoRentals)<self.minPromos:
            return 0

        promoSums = 0
        for rental in self.promoRentals:
            promoSums = promoSums + rental.getTotalCost()
        return promoSums

    def addRental(self, aRental):
        if len(self.promoRentals)<self.maxPromos:
            self.promoRentals.append(aRental)
        else:
            raise ValueError('Maximum rental items reached')

class CompositeFamilyPromotionTest(unittest.TestCase):
    def test_addRentalAndTotalCost(self):
        promo = CompositeFamilyPromotion()
        rental = RentByDay(1)
        promo.addRental(rental)
        rental = RentByDay(1)
        promo.addRental(rental)
        rental = RentByDay(1)
        promo.addRental(rental)
        rental = RentByDay(1)
        promo.addRental(rental)
        self.assertEqual(promo.getTotalCost(),80)


    def test_getTotalCostWhenLessThanMin(self):
        promo = CompositeFamilyPromotion()
        rental = RentByDay(1)
        promo.addRental(rental)
        self.assertEqual(promo.getTotalCost(),0)


if __name__ == '__main__':
    unittest.main()
